testsuite/testcases/src/validator_join_leave_test.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{LoadDestination, NetworkLoadTest};
+use aptos::account::create::DEFAULT_FUNDED_COINS;
+use aptos_logger::info;
+use aptos_sdk::crypto::ed25519::Ed25519PrivateKey;
+use aptos_sdk::crypto::PrivateKey;
+use forge::{
+    reconfig, NetworkContext, NetworkTest, NodeExt, Result, Swarm, SwarmExt, Test, FORGE_KEY_SEED,
+};
+
+use aptos_keygen::KeyGen;
+
+use aptos::test::CliTestFramework;
+use aptos_types::account_address::AccountAddress;
+use aptos_types::transaction::authenticator::AuthenticationKey;
+use std::time::Duration;
+use tokio::runtime::Runtime;
+
+const MAX_NODE_LAG_SECS: u64 = 360;
+
+pub struct ValidatorJoinLeaveTest;
+
+impl Test for ValidatorJoinLeaveTest {
+    fn name(&self) -> &'static str {
+        "validator join and leave sets"
+    }
+}
+
+impl NetworkLoadTest for ValidatorJoinLeaveTest {
+    fn setup(&self, _ctx: &mut NetworkContext) -> Result<LoadDestination> {
+        Ok(LoadDestination::AllValidators)
+    }
+
+    fn test(&self, swarm: &mut dyn Swarm, duration: Duration) -> Result<()> {
+        // Verify we have at least 7 validators (i.e., 3f+1, where f is 2)
+        // so we can lose 2 validators but still make progress.
+        let all_validators = swarm.validators().map(|v| v.peer_id()).collect::<Vec<_>>();
+        let num_validators = all_validators.len();
+        if num_validators < 7 {
+            return Err(anyhow::format_err!(
+                "ValidatorSet leaving and rejoining test require at least 7 validators! Given: {:?}.",
+                num_validators
+            ));
+        }
+
+        let faucet_endpoint: reqwest::Url = "http://localhost:8081".parse().unwrap();
+        // Connect the operator tool to the node's JSON RPC API
+        let rest_client = swarm.validators().next().unwrap().rest_client();
+        let transaction_factory = swarm.chain_info().transaction_factory();
+        let runtime = Runtime::new().unwrap();
+
+        let mut cli = runtime.block_on(async {
+            CliTestFramework::new(
+                swarm.validators().next().unwrap().rest_api_endpoint(),
```

Comment:
use rest_client.clone() here?

Reply:
The parameter is a URL, not a rest client.